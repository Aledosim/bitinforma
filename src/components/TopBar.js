import React, { useContext } from 'react'

import { CurrencyContext } from '../contexts/CurrencyContext'

import { configs } from '../configs'
import styles from '../styles/components/TopBar.module.css'

import logo from "../images/logo.svg"

export function search(setCurrency) {
  const searchInner = (event) => {
    const searchField  = document.getElementById('searchField')
    const text = searchField.value.toLowerCase()

    if (Object.keys(configs.currencies).includes(text)) {
      setCurrency(text)
      searchField.value = ''
    } else {
      searchField.value = ''
      searchField.placeholder = 'Termo inv√°lido'
    }
    event.preventDefault()
  }

  return searchInner
}

export default function TopBar(){

  const {
    setCurrency
  } = useContext(CurrencyContext)
        // <input
        //   data-cy='searchField'
        //   id='searchField'
        //   type='text'
        //   placeholder='Buscar por uma moeda'
        // />

        // <img
        //   data-cy='searchButton'
        //   src={search_icon}
        //   alt='search button'
        //   onClick={search(setCurrency)}
        // />
  return(
    <header data-cy='topbar' className={styles.topBar}>
      <div>
        <img className={styles.logo} src={logo} alt='logo' />
      </div>
      <form
        onSubmit={search(setCurrency)}
        className={styles.searchFieldContainer}
      >
        <input
          type='search'
          autoComplete='on'
          autoSave='true'
          placeholder='Buscar por uma moeda'
          id='searchField'
          data-cy='searchField'
        />
        <button
          type='submit'
          className="fas fa-search"
          data-cy='searchButton'
        />
      </form>
    </header>
  );
};
