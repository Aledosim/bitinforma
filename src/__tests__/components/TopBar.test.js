import { enableFetchMocks  } from 'jest-fetch-mock'

enableFetchMocks()
import React from 'react'
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react'

import { tickerResponse, summaryResponse } from '../fixtures.json'
import CurrencyProvider from '../../contexts/CurrencyContext'

import TopBar, { search } from '../../components/TopBar'

function testRender() {
    return render(
      <CurrencyProvider>
        <TopBar />
      </CurrencyProvider>
    )
}

describe('search tests', () => {

  it('returns a function', () => {
    const func = search(jest.fn())

    expect(typeof func).toEqual('function')
  })

  it('gets the correct input', () => {
    const event = jest.fn()
    const mockPreventDefault = jest.fn()
    event.preventDefault = mockPreventDefault

    document.getElementById = jest.fn(() => HTMLInputElement)
    HTMLInputElement.value = ''

    const func = search(jest.fn())
    func(event)

    expect(document.getElementById).toHaveBeenCalledWith('searchField')
  })

  it('search for lowercase only', () => {
    const event = jest.fn()
    const mockPreventDefault = jest.fn()
    event.preventDefault = mockPreventDefault

    HTMLInputElement.value = 'ETH'
    document.getElementById = jest.fn(() => HTMLInputElement)
    const setCurrency = jest.fn()

    const func = search(setCurrency)
    func(event)

    expect(setCurrency).toHaveBeenCalledWith('eth')

  })

  it('calls setCurrency for a valid value and reset searchField', () => {
    const event = jest.fn()
    const mockPreventDefault = jest.fn()
    event.preventDefault = mockPreventDefault

    const setCurrency = jest.fn()
    HTMLInputElement.value = 'eth'
    document.getElementById = jest.fn(() => HTMLInputElement)

    const func = search(setCurrency)
    func(event)

    expect(setCurrency).toHaveBeenCalledWith('eth')
    expect(HTMLInputElement.value).toEqual('')
  })

  it('wont call setCurrency for a invalid value and show a message', () => {
    const event = jest.fn()
    const mockPreventDefault = jest.fn()
    event.preventDefault = mockPreventDefault

    const setCurrency = jest.fn()

    document.getElementById = jest.fn(() => HTMLInputElement)
    HTMLInputElement.value = 'invalid'

    const func = search(setCurrency)
    func(event)

    expect(setCurrency).not.toHaveBeenCalled()
    expect(HTMLInputElement.placeholder).toEqual('Termo invÃ¡lido')
    expect(HTMLInputElement.value).toEqual('')
  })

  it('calls event.disableDefault method to prevent page reload', () => {
    const event = jest.fn()
    const mockPreventDefault = jest.fn()
    event.preventDefault = mockPreventDefault

    HTMLInputElement.value = ''
    document.getElementById = jest.fn(() => HTMLInputElement)

    const func = search()
    func(event)

    expect(mockPreventDefault).toHaveBeenCalled()
  });
})

describe('<TopBar /> tests', () => {
  beforeEach(() => {
    fetch.resetMocks()

    fetch
      .mockResponse(req => {
        if (/.*\/ticker\/.*/.test(req.url)) {
          return Promise.resolve({ body: JSON.stringify(tickerResponse) })

        } else if (/.*\/day-summary\/.*/.test(req.url)) {
          return Promise.resolve({ body: JSON.stringify(summaryResponse) })

        }
      })
  })

  it('call search when search button is clicked', async () => {
    const mockInnerSearch = jest.fn()
    const mockSearch = jest.fn(() => mockInnerSearch)
    TopBar.__Rewire__('search', mockSearch)
    // window.HTMLFormElement.prototype.submit = () => mockSearch

    testRender()

    fireEvent.submit(screen.getByRole('button'))
    expect(mockInnerSearch).toHaveBeenCalled()

    TopBar.__ResetDependency__('search')

    // Check if the fetch data is rendered
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(2)
    })
  })

  it('renders without crashing', async () => {
    const tree = testRender()

    expect(tree).toMatchSnapshot()

    // Check if the fetch data is rendered
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(2)
    })
  })


})
