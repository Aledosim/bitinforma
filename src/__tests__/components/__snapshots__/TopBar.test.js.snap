// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> tests renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="topBar"
        data-cy="topbar"
      >
        <div>
          <img
            alt="logo"
            class="logo"
            src="logo.svg"
          />
        </div>
        <form
          class="searchFieldContainer"
        >
          <input
            autocomplete="on"
            autosave="true"
            data-cy="searchField"
            id="searchField"
            placeholder="Buscar por uma moeda"
            type="search"
          />
          <button
            class="fas fa-search"
            data-cy="searchButton"
            type="submit"
          />
        </form>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="topBar"
      data-cy="topbar"
    >
      <div>
        <img
          alt="logo"
          class="logo"
          src="logo.svg"
        />
      </div>
      <form
        class="searchFieldContainer"
      >
        <input
          autocomplete="on"
          autosave="true"
          data-cy="searchField"
          id="searchField"
          placeholder="Buscar por uma moeda"
          type="search"
        />
        <button
          class="fas fa-search"
          data-cy="searchButton"
          type="submit"
        />
      </form>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
